UKHomeOffice/brp_app
# Add redis session storage

### Testing

Prerequisites: 
- redis server installed
- `npm install`

Steps to run: 
- Open a new terminal and run `redis-server`.
- Assuming your redis server is running successfully in the background on default port 6379, start the app with  `npm start`
- Navigate the frontend as normal and confirm that the date entered persists to the check details pages.

Not currently sure how to verify the data persists between instances of the app or what to do about clustering redis.

### PR explained
- install express session
- install redis client
- update config
- make config mutable with env vars 
- document env vars

- app.js
    - create redis client with port and host
    - assign redis client instance to `connect-redis` (new RedisStore) session store and set expiry to 15 mins (ttl)
    - export cookie/session middleware

Cookie support added for to avoid error for undefined `req.cookies`:
`node_modules/hmpo-form-wizard/lib/middleware/check-session.js:4:28`

Although `controller.options.checkSession` can be set to false to avoid the error . It seemed like very little overhead to add cookie support once as opposed to potential needing to set `controller.options.checkSession` to `false` for every controller. 

Therefore we have the following middleware for handling cookies: 

```js 
function secureCookies(req, res, next) {
  var cookie = res.cookie.bind(res);
  res.cookie = function cookieHandler(name, value, options) {
    options = options || {};
    options.secure = (req.protocol === 'https');
    options.httpOnly = true;
    options.path = '/';
    cookie(name, value, options);
  };
  next();
}
```

And for initialising the session with the redis store and cookie handling:
```js 
function initSession(req, res, next) {
  session({
    store: redisStore,
    cookie: {
      secure: (req.protocol === 'https')
    },
    key: 'hmbrp.sid',
    secret: config.session.secret,
    resave: true,
    saveUninitialized: true
  })(req, res, next);
}
```
Finally the following exports and adds the functions to the set of middleware
```js 
module.exports = [
  require('cookie-parser')(config.session.secret),
  secureCookies,
  initSession
];
```
